packages/hub/services/scheduled-payments/fetcher.ts

```diff
@@ -19,44 +21,74 @@ export default class ScheduledPaymentsFetcherService {
-                gte: nowUtc,
-              },
-              scheduledPaymentAttempts: {
-                none: {
-                  status: 'inProgress',
-                },
-              },
-            },
-          },
-        ],
-      },
-    });
+    let results: [{ id: string; started_at: Date }] = await prisma.$queryRaw`SELECT 
+        "scheduled_payments"."id" AS "id",
+        "last_failed_payment_attempts"."started_at"
+      FROM 
+        "scheduled_payments"
+      LEFT JOIN (SELECT "scheduled_payment_attempts"."scheduled_payment_id",
+                  MAX("scheduled_payment_attempts"."started_at") AS "started_at"
+            FROM "scheduled_payment_attempts"
+            WHERE "scheduled_payment_attempts"."status" = 'failed'
+            GROUP BY "scheduled_payment_id") AS "last_failed_payment_attempts"
+      ON "last_failed_payment_attempts"."scheduled_payment_id" = "scheduled_payments"."id"
+      WHERE 
+        (
+          "scheduled_payments"."canceled_at" IS NULL 
+          AND "scheduled_payments"."creation_block_number" > 0 
+          AND "scheduled_payments"."pay_at" > ${startOfDay(subDays(nowUtc, this.validForDays))}
+          AND "scheduled_payments"."pay_at" <= ${nowUtc}
+          AND (
+            (
+              "scheduled_payments"."recurring_day_of_month" IS NULL 
+              AND (
+                "scheduled_payments"."id"
+              ) NOT IN (
+                SELECT 
+                  "t0"."id" 
+                FROM 
+                  "scheduled_payments" AS "t0" 
+                  INNER JOIN "scheduled_payment_attempts" AS "j0" ON ("j0"."scheduled_payment_id") = ("t0"."id") 
+                WHERE 
+                  (
+                    (
+                      "j0"."status" = 'succeeded'
+                      OR "j0"."status" = 'in_progress'
+                    ) 
+                    AND "t0"."id" IS NOT NULL
+                  )
+              )
+            ) 
+            OR (
+              "scheduled_payments"."recurring_day_of_month" > 0 
+              AND "scheduled_payments"."recurring_until" >= ${nowUtc}
+              AND (
+                "scheduled_payments"."id"
+              ) NOT IN (
+                SELECT 
+                  "t0"."id" 
+                FROM 
+                  "scheduled_payments" AS "t0" 
+                  INNER JOIN "scheduled_payment_attempts" AS "j0" ON ("j0"."scheduled_payment_id") = ("t0"."id") 
+                WHERE 
+                  (
+                    "j0"."status" = 'in_progress'
+                    AND "t0"."id" IS NOT NULL
+                  )
+              )
+            )
+          )
+        )
+        ORDER BY CASE WHEN  "last_failed_payment_attempts"."started_at" IS NULL THEN 0 ELSE 1 END
+        LIMIT 10;`;
+
+    // Retrieve scheduledPayment prisma object based on filtered id
+    return prisma.scheduledPayment.findMany({ where: { id: { in: results.map((result) => result.id) } } });
```

Comment:
This could be done in a subquery together with the query above.

Reply:
Is there a way to automatically map the DB result to the Prisma object? I did this because the naming fields of `scheduled_payments` table and `scheduledPayment` Prisma object are different. For example, `chain_id` and `chainId`. Another solution is to use `AS` in the raw query so we can rename the field, but I think it's not easy to maintain because if we add another field or rename the existing field to `scheduled_payments` table, we always need to update that raw query.

Reply:
Ok I see the issue. It's a pity how unflexible Prisma is for nontrivial queries.